tests/test_site_apportionment/test_site_apportionment.py

```diff
@@ -1,18 +1,431 @@
+        input_df = create_exp_percent_test_output_df
+        exp_to_apportion_df = self.create_exp_to_apportion_output()
+        exp_df_out = self.create_exp_remaining_output()
+
+        imp_markers_to_keep: list = ["R", "TMI", "CF", "MoR", "constructed"]
+        result_to_apportion, result_df_out = split_sites_df(
+            input_df, 
+            imp_markers_to_keep
+        )
+
+        assert_frame_equal(result_to_apportion.reset_index(drop=True), exp_to_apportion_df)
+
+        assert_frame_equal(result_df_out.reset_index(drop=True), exp_df_out)
+
+
+class TestDeduplicateCodeValues():
+    """Tests for the deduplicate_codes_values function."""
+
+    def create_input_df(self):
+        """Create an input dataframe for the test."""
+        input_columns = [
+            "reference",
+            "period",
+            "200",
+            "201",
+            "pg_numeric",
+            "211",
+            "251",
+        ]
+            
+        data = [
+            [111, "2020", "C", "A", 42, 5000.0, "Yes"],
+            [111, "2020", "C", "A", 42, 777.0, "No"],
+            [111, "2020", "D", "B", 37, 300.0, "Yes"],
+            [222, "2020", "C", "A", 61, 1000.0, "Yes"],
+            [222, "2020", "C", "A", 63, 500.0, "No"],
+        ]   
+
+        input_df = pandasDF(data=data, columns=input_columns)
+        return input_df
+        
+    def create_exp_output_df(self):
+        """Create an input dataframe for the test."""
+        exp_output_columns = [
+            "reference",
+            "period",
+            "200",
+            "201",
+            "pg_numeric",
+            "211",
+            "251",
+        ]
+            
+        data = [
+            [111, "2020", "C", "A", 42, 5777.0, "Yes"],
+            [111, "2020", "D", "B", 37, 300.0, "Yes"],
+            [222, "2020", "C", "A", 61, 1000.0, "Yes"],
+            [222, "2020", "C", "A", 63, 500.0, "No"],
+        ]   
+
+        exp_output_df = pandasDF(data=data, columns=exp_output_columns)
+        return exp_output_df
+    
+    def test_deduplicate_codes_values(self):
+        """Test for the deduplicate_codes_values function."""
+        input_df = self.create_input_df()
+        exp_df = self.create_exp_output_df()
+
+        result_df = deduplicate_codes_values(
+            input_df,
+            ["reference", "period", "200", "201", "pg_numeric"],
+            ["211"],
+            ["251"]
+        )
+
+        assert_frame_equal(result_df.reset_index(drop=True), exp_df)
+    
+
+class TestCalcWeightsForSites():
+    """Tests for the calc_weights_for_sites function."""
+    def create_input_df(self):
+        """Create an input dataframe for the test."""
+        input_columns = [
+            "reference",
+            "period",
+            "instance",
+            "601",
+            "602",
+            "postcodes_harmonised",
+        ]
+            
+        data = [
+            [111, "2020", 1, "AB15 3GU", 0, "AB15 3GU"],
```

Comment:
ok, I've fixed this. (The reason I didn't have it at first is because that line of code filling na with zero shouldn't be necessary as there shouldn't be any nulls in that dataframe. But just as well to leave it in, and if it is in, it should be tested)